README.md
# The PetFinder Group

A Command Line Tool created to help users find cats up for adoption. 

Written as part two of an interview with The PathFinder Group by Charlie Malachinski. 

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.


### Installation

Make sure that you have all the [requirements](requirements.txt) installed. 

A step by step guide that will tell you how to get the development environment up and running.

```
$ Run 'server-run.bat' to connect JSON to endpoint.
$ Open Command Prompt or Terminal and navigate to install location. 
$ Use command 'python main.py --help' to see list of all commands. 
```

## Usage

A few examples of useful commands and/or tasks.

```
$ python main.py list --> shows a list of all available cats 
$ python main.py age --> shows a filter list of cats based on age argument
$ python main.py gender --save --> shows a filter list of cats based on gender argument and prompts user to save results as a CSV
```


## Additional Documentation and Acknowledgments

* Implements [ JSON-SERVER ](https://www.npmjs.com/package/json-server)to mock test rest server. 
* Implements [Click](https://click.palletsprojects.com/en/8.1.x/) for commands and options. 
* Implements [Requests](https://requests.readthedocs.io/en/latest/) for HTTP methods.
* Implements [Pandas](https://pandas.pydata.org/) for JSON, CSV, and DataFrame compatibility. 


release-notes.txt
#----------------------------------------------------------------------------------------------------#

The PetFinder Group. Version 1.0.0, 10/28/22

New Features:
    - Application displays original cat availability data to the users. 
	- Application allows users to filter through data based on the following criteria: Age, Color, Declawed, Gender, Name. 
	- Application displays filtered data in CSV format for better readability.
	- Application allows users to export both original and filtered data to a new CSV file. 
	
Known Issues:
	- N/A

Expected Next Version:
	- Ability to add new cats to the list (using HTTPS PUTS methods)

Links:
	- GitHub Repository: https://github.com/charliemalach/-----

#----------------------------------------------------------------------------------------------------#

src/methods.py
import json
import click
import requests
import pandas as pd
import sys
from requests.exceptions import ConnectionError

# The PetFinder Group
# @author Charlie Malachinski
# @version 1.0.0
# @since 10-28-22
# This class file is used to host the methods that are used by the main class file. 

# sets the localhost server as the endpoint for the program
URL = "http://localhost:3000/cats"

class methods:
    # function to return the data with filter arguments  
    def filter_data(selection = None):
        try: 
            request = requests.get(URL, params=selection)
            return request.content
        # i was running into an error when the server was not running, so now there is a reminder for users to run the server before executing commands 
        except ConnectionError:
                methods.print_bar()
                print("ERROR! Could not connect to JSON SERVER. Please run 'server-run.bat' and try again.")
                methods.print_bar()
                sys.exit(3)
            
    # function to filter the entry by the user. takes care of the case that user enters CAPS vs lowercase.
    def format_text(entry):
        entry = entry.lower().capitalize()
        return entry

    # function to print bars just for the aesthetics of the command prompt 
    def print_bar():
        print("---------------------------------------------------------------------------------------------")

    #function to convert the json response to a CSV
    def convert_csv(response):
        output = json.dumps(response.decode('utf-8'))
        json_data = json.loads(output)
        df = pd.read_json(json_data)
        return df 

    # displays the filtered results as csv (i think this looks MUCH better)
    def display_results(response):
        methods.print_bar()
        print("Here are your results: ")
        methods.print_bar()
        df = methods.convert_csv(response)
        if df.empty:
            click.echo('Error: No results matched your criteria. Please try again with different criteria.')
        else:
            click.echo(df.to_csv(index=False))
            
    # function to save the search response to a CSV
    def save_csv(response):
        df = methods.convert_csv(response)
        file = input("Please name new CSV output:")
        df.to_csv('../results/'+ file + '.csv', index=False)
        click.echo("Success! Your results have been saved to " + file + ".csv.")

    #function to print results and save
    def save_results(response):
        methods.display_results(response)
        methods.save_csv(response)

    # final function to save and print or just print results  
    def results(response, save = None):
        if not save:
            methods.display_results(response)
            methods.print_bar()
        else:
            methods.save_results(response)
            methods.print_bar()

src/main.py
import click
from methods import methods

# The PetFinder Group
# @author Charlie Malachinski
# @version 1.0.0
# @since 10-28-22
# This file is used as the main class 'driver' file. Has mostly the Click commands and their respective options. Uses methods from 'methods.py'

# creates a group using Click and adds help option to the group 
@click.group()
@click.help_option('--help', help='Displays this screen and all available commands.')

# sorry for the awkward formatting in the text, I thought it would bea nice touch for the UI d:^)~
def main():
    """------------------------------ \n
\n### The PetFinder Group. ###  \n \n
\n - We are dedicated to finding forever homes for beautiful cats. \n
\n - Here are some of the basic commands for using our program. \n
------------------------------"""
    pass

# this command shows the list of cats filtered by AGE with the option to save the filtered list of cats to a CSV
@click.command()
@click.option("--age", prompt="Enter the desired age", help="The age of the cat", type=int)
@click.option("--save", help="Saves the data to a new path", is_flag=True) #option to save results to CSV
def age(age, save):
    """Simple command that filters cats based on AGE."""
    payload = {"age": age}
    result = methods.filter_data(payload) #filters through data based on age
    methods.results(result, save)

# this command shows the list of cats filtered by COLOR with the option to save the filtered list of cats to a CSV
@click.command()
@click.option("--color", prompt="Enter the desired color", help="The color of the cat")
@click.option("--save", help="Saves the data to a new path", is_flag=True) #option to save results to CSV
def color(color, save):
    """Simple command that filters cats based on COLOR."""
    color = methods.format_text(color) #formats the color 
    payload = {"color": color}
    result = methods.filter_data(payload) #filters through data based on color
    methods.results(result, save)

# this command shows the list of cats filtered by being DECLAWED with the option to save the filtered list of cats to a CSV
@click.command()
@click.option("--declaw", prompt="Do you want a cat that is declawed? (Y/N)", help="Determines if the cat is declawed or not.")
@click.option("--save", help="Saves the data to a new path", is_flag=True) #option to save results to CSV
def declawed(declaw, save):
    """Simple command that filters cats based on being DECLAWED."""
    declaw = methods.format_text(declaw) #formats the text for declaw 
    payload = {"declawed": declaw}
    result = methods.filter_data(payload) #filters through data based on cat being declawed 
    methods.results(result, save)

# this command shows the list of cats filtered by GENDER with the option to save the filtered list of cats to a CSV
@click.command()
@click.option("--gender", prompt="Enter the desired gender (M/F)", help="The gender of the cat")
@click.option("--save", help="Saves the data to a new path", is_flag=True) #option to save results to CSV
def gender(gender, save):
    """Simple command that filters cats based on GENDER."""
    gender = methods.format_text(gender) #formats the gender 
    payload = {"gender": gender}
    result = methods.filter_data(payload) #filters through data based on gender
    methods.results(result, save)

# this command shows the list of cats filtered by NAME with the option to save the filtered list of cats to a CSV
@click.command()
@click.option("--name", prompt="Enter the desired name", help="The name of the cat")
@click.option("--save", help="Saves the data to a new path", is_flag=True) #option to save results to CSV
def name(name, save):
    """Simple command that filters cats based on NAME."""
    name = methods.format_text(name) #formats the name 
    payload = {"name": name}
    result = methods.filter_data(payload) #filters through data based on name
    methods.results(result, save)

# this command shows the original list of ALL cats with the option to save ALL cats as a CSV
@click.command()
@click.option("--save", help="Saves the data to a new path", is_flag=True)
def list(save):
    """Simple command that lists ALL original cats."""
    result = methods.filter_data() #fetches original data
    methods.results(result, save)

# add all commands to the CLI
main.add_command(age)
main.add_command(color)
main.add_command(declawed)
main.add_command(gender)
main.add_command(list)
main.add_command(name)

# driver the cause main to run
if __name__ == "__main__":
    main()